### PR TITLE
[FEATURE] Add Baby Bear Field

### PR DESCRIPTION
**Pull request by enpsi20481**
- [x] Add Baby Bear field to `arith` package as an import from Plonky3
- [x] Add the degree 3 extension of Baby Bear
- [x] Add the degree 4 extension of Baby Bear
- [x] Add SIMD support
    - [x] AVX512
    - [x] Neon

### CODE REVIEW

**Comment by enpsi20481** [comment]
Unsure how to deal with the `avx512f` target feature. The x86 build imports its BabyBear SIMD implementation from Plonky3, where it is gated behind `"target-feature=+avx512f"` so `arith` only builds on x86 with that feature activated, e.g. `RUST_FLAGS="-C target-feature=+avx512f"`.

It's cumbersome to always pass in that flag so I added a `.cargo/config.toml` to turn it on, but then `arith` will not build on an `aarch` target. Currently I've left the `config.toml` file there with that line commented out so that it is easy to toggle the feature depending on the platform, but I am not sure this is the best solution

File: arith/Cargo.toml [hunk 1]
```diff
@@ -7,6 +7,10 @@ edition = "2021"
 ark-std.workspace = true
 halo2curves.workspace = true
 log.workspace = true
+p3-baby-bear = { git = "https://github.com/Plonky3/Plonky3.git", version = "0.1.0", features = ["nightly-features"] }
```
Comment (zhenfeizhang) [inline arith/Cargo.toml]: nits: please move this to workspace for consistency

File: arith/src/extension_field/baby_bear_ext3.rs [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+// = a0 b0 + w * (a1 b2 + a2 b1)
+//   + {(a0 b1 + a1 b0) + w * a2 b2} x
+//   + {(a0 b2 + a1 b1 + a2 b0)} x^2
+#[inline(always)]
+fn mul_internal(a: &BabyBearExt3, b: &BabyBearExt3) -> BabyBearExt3 {
+    let w = BabyBear::new(BabyBearExt3::W);
+    let a = a.v;
+    let b = b.v;
+    let mut res = [BabyBear::default(); 3];
+    res[0] = a[0] * b[0] + w * (a[1] * b[2] + a[2] * b[1]);
```
Comment (zhenfeizhang) [inline arith/src/extension_field/baby_bear_ext3.rs]: nits: `W` is two so we should use `double()` for performance

File: arith/src/extension_field/baby_bear_ext3.rs [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+    res[1] = (a[0] * b[1] + a[1] * b[0]) + w * a[2] * b[2];
+    res[2] = a[0] * b[2] + a[1] * b[1] + a[2] * b[0];
+    BabyBearExt3 { v: res }
+}
+
+#[inline(always)]
+fn square_internal(a: &[BabyBear; 3]) -> [BabyBear; 3] {
+    let w = BabyBear::new(BabyBearExt3::W);
+    let mut res = [BabyBear::default(); 3];
+    res[0] = a[0].square() + w * (a[1] * a[2]).double();
```
Comment (zhenfeizhang) [inline arith/src/extension_field/baby_bear_ext3.rs]: same. avoid mul by w

File: arith/src/extension_field/baby_bear_ext3x16.rs [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+        //
+        // (a0 + a1 x + a2 x^2) * (b0 + b1 x + b2 x^2)
+        // = a0 b0 + (a0 b1 + a1 b0) x + (a0 b2 + a1 b1 + a2 b0) x^2 + (a1 b2 + a2 b1) x^3 + a2 b2 x^4
+        // = a0 b0 + w * (a1 b2 + a2 b1)
+        //   + {(a0 b1 + a1 b0) + w * a2 b2} x
+        //   + {(a0 b2 + a1 b1 + a2 b0)} x^2
+
+        let w = BabyBear::from(BabyBearExt3x16::W);
+        let mut res = [BabyBearx16::ZERO; 3];
+        res[0] = self.v[0] * rhs.v[0] + (self.v[1] * rhs.v[2] + self.v[2] * rhs.v[1]) * w;
```
Comment (zhenfeizhang) [inline arith/src/extension_field/baby_bear_ext3x16.rs]: - use `square()` for `mul`
- use `double()` for `* w`

File: arith/src/extension_field/baby_bear_ext3x16.rs [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+    // (a0 + a1 x + a2 x^2) * (b0 + b1 x + b2 x^2)
+    // = a0 b0 + (a0 b1 + a1 b0) x + (a0 b2 + a1 b1 + a2 b0) x^2 + (a1 b2 + a2 b1) x^3 + a2 b2 x^4
+    // = a0 b0 + w * (a1 b2 + a2 b1)
+    //   + {(a0 b1 + a1 b0) + w * a2 b2} x
+    //   + {(a0 b2 + a1 b1 + a2 b0)} x^2
+    let a = &a.v;
+    let b = &b.v;
+    let mut res = [BabyBearx16::default(); 3];
+    let w = BabyBear::from(BabyBearExt3x16::W);
+    res[0] = a[0] * b[0] + (a[1] * b[2] + a[2] * b[1]) * w;
```
Comment (zhenfeizhang) [inline arith/src/extension_field/baby_bear_ext3x16.rs]: double

File: arith/Cargo.toml (hunk 1 above)
Comment (enpsi20481) [reply arith/Cargo.toml]: done

File: arith/src/extension_field/baby_bear_ext3.rs (hunk 2 above)
Comment (enpsi20481) [reply arith/src/extension_field/baby_bear_ext3.rs]: good point, done

File: arith/src/extension_field/baby_bear_ext3.rs (hunk 3 above)
Comment (enpsi20481) [reply arith/src/extension_field/baby_bear_ext3.rs]: done

File: arith/src/extension_field/baby_bear_ext3x16.rs (hunk 4 above)
Comment (enpsi20481) [reply arith/src/extension_field/baby_bear_ext3x16.rs]: Used `double()`, not sure where to use `square` (?)

File: arith/src/extension_field/baby_bear_ext3x16.rs (hunk 5 above)
Comment (enpsi20481) [reply arith/src/extension_field/baby_bear_ext3x16.rs]: done

**Pull request closed without merging**